src/extension-runners/chromium.js

```diff
@@ -127,8 +127,58 @@ export class ChromiumExtensionRunner {
       chromeFlags.push(...this.params.args);
     }
 
-    if (this.params.chromiumProfile) {
-      chromeFlags.push(`--user-data-dir=${this.params.chromiumProfile}`);
+    function isUserDataDir(dirCandidate: string): boolean {
+      const localStateCandidate = path.join(dirCandidate, 'Local State');
+      const defaultCandidate = path.join(dirCandidate, 'Default');
+      // Local State and Default are typical for the user-data-dir
+      return fs.pathExistsSync(localStateCandidate)
+        && fs.pathExistsSync(defaultCandidate);
+
+    }
+
+    function isProfileDir(dirCandidate: string): boolean {
+      const securePreferencesCandidate = path.join(
+        dirCandidate,
+        'Secure Preferences');
+      //Secure Preferences is typical for a profile dir
+      return fs.pathExistsSync(securePreferencesCandidate);
+    }
+
+    let userDataDir = false;
+    const chromiumProfile = this.params.chromiumProfile;
+    if (chromiumProfile) {
+      const isProfileDirAndNotUserData =
+        fs.pathExistsSync(chromiumProfile)
+        && isProfileDir(chromiumProfile)
+        && !isUserDataDir(chromiumProfile);
+
+      if (isProfileDirAndNotUserData) {
+        const pathParts = path.parse(chromiumProfile);
+        const profileDirName = pathParts.base;
+        if (this.params.keepProfileChanges) {
+          // now check if the parent dir is a user-data-dir
+          if (!isUserDataDir(pathParts.dir)) {
+            throw new Error('The profile you provided is not in a ' +
+              'user-data-dir. The changes cannot be kept. Please either ' +
+              'remove --keep-profile-changes or use a profile in a ' +
+              'user-data-dir directory');
+          }
+          userDataDir = pathParts.dir;
+
+        } else {
+          // the user provided an existing profile directory but doesn't want
+          // the changes to be kept. we copy this directory to a temporary
+          // user data dir.
+          userDataDir = Launcher.prototype.makeTmpDir();
```

Comment:
Don't use undocumented internal implementation details from an external library.

The `web-ext` project itself already has an utility to create temporary directories here: https://github.com/mozilla/web-ext/blob/9315b781ec73eead946c07551a24a413ddb1338a/src/util/temp-dir.js

An example of usage is here: https://github.com/mozilla/web-ext/tree/9315b781ec73eead946c07551a24a413ddb1338a/src/extension-runners

Reply:
I switched to TempDir, but with TempDir you cannot just get a temp path without creating the directory itself. I have to delete it and rewrite it after, which is quite ugly. My method would like to just get a path, and create the actual dir separately...